proposals/0540-vision-for-layout-conformance-parity.md

```diff
@@ -0,0 +1,238 @@
+
+# Vision for Layout Conformance/Parity
+
+React Native models its layout and styling support after web browsers, offering a subset of CSS features. Some CSS properties like color are mapped to functionality in the backing UI framework. Sizing, positioning, and general layout is handled by the Yoga layout engine. This intersection gives web developers a familiar experience when working in React Native, but also behaves similarly enough to web to enable code-sharing, as supported by projects like react-native-web.
+
+There are gaps in what React Native supports, with “CSS Features” being the [#1 request](https://github.com/react-native-community/discussions-and-proposals/discussions/528?sort=top#discussioncomment-3886000) in a recent survey to the community over the state of React Native. This intertwines with asks for us to invest in Yoga, which has been [inactive](https://github.com/facebook/yoga/issues/1151). Flexbox support in React Native has had limitations akin to targeting [Internet Explorer 9](http://caniuse.com/?search=column-gap). We also hear a desire for more consistent behavior of existing behaviors with web. This document outlines a set of problems, strategies, and solutions, for building out the intertwined problems of layout conformance, and layout parity.
+
+## Where are the differences?
+
+Differences between layout in RN vs web can roughly be categorized into:
+
+1. **Algorithm Differences:** Given the same styles, normalized for defaults, React Native may still produce a different layout from web browsers. An example of this is a Yoga flag `UseLegacyStretchBehaviour`, where for backwards compatibility, React Native today intentionally diverges from web spec.
+2. **Different default behavior:** A new box in react-native is given different [default style properties](https://reactnative.dev/docs/flexbox#flex) than a box on web. E.g. items flow in a column by default instead of a row. Standalone Yoga allows controlling some of these via a `UseWebDefaults` configuration flag, and existing frameworks like `react-native-web` bridge the gap by setting React Native’s default [styles](https://github.com/necolas/react-native-web/blob/dee258ae02188befca4192ea502d4fbebc478341/packages/react-native-web/src/exports/View/index.js#L157) on each View.
+3. **Capabilities:** Browser-makers continually add support for new layout capabilities to CSS. This includes new layout systems like CSS grid, but also continual [revisions](https://www.w3.org/TR/css-align-3/) standards which form the basis of Flexbox (e.g. for row-gap and column-gap). 
+
+Where new capabilities are additive, changing default behavior, or existing layout algorithms would affect the way existing components renders. Even if the new behavior is correct, we must be conscious to preserve compatibility with the vast ecosystem of existing React Native components.
+
+## Strict layout
```

Comment:
@NickGerleman 
I would like to propose an alternative to this. This proposal is inspired by React's handling of the deprecated lifecycle methods like `componentWillReceiveProps` by renaming the method to `UNSAFE_componentWillReceiveProps` and providing a codemod that automatically applied the migration.

Instead of a `StrictLayout` component that affects the entire tree below it, I would like to propose that the default behaviour of `View` is changed to the new web compatible behaviour, and that a new `LegacyView` component (or an alternatively a `useLegacyLayout` prop on the `View` component) is introduced which opts back into the legacy layout algorithm for that view only. Additionally a codemod would be provided that would automatically convert all `View`s to `LegacyView`.

If you wanted to be more sure that people didn't forget to run this codemod, then the `View` component could also be renamed to some other name (making not running the codemod a build error). Perhaps in order to compensate for View having a less convenient name, new `Row`, `Column` and `Grid` components could be introduced which function like `View` but with the layout algorithm and orientation already set.